### ISSUE TITLE
Clean up ugni buffered operation implementation

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#12067]
The current buffering code is hard to read and not very maintainable. We had
originally deferred cleaning it up in order to make progress and demonstrate
performance, but the current code needs to be cleaned up before we start
looking at moving it into a common shim (so we can use it with different comm
layers)

WIP branch exists https://github.com/ronawho/chapel/tree/cleanup-amo-buff.
Cleanup is mostly mostly done for buffered AMOs, but needs to be done for
buffered GETs as well.

Related discussion for cleanup:
 - https://github.com/chapel-lang/chapel/pull/10702#discussion_r208717023
 - https://github.com/chapel-lang/chapel/pull/11338#discussion_r234037842
